tests/functional_tests/ffront_tests/test_program.py

```diff
@@ -0,0 +1,291 @@
+# GT4Py Project - GridTools Framework
+#
+# Copyright (c) 2014-2021, ETH Zurich
+# All rights reserved.
+#
+# This file is part of the GT4Py project and the GridTools framework.
+# GT4Py is free software: you can redistribute it and/or modify it under
+# the terms of the GNU General Public License as published by the
+# Free Software Foundation, either version 3 of the License, or any later
+# version. See the LICENSE.txt file at the top-level directory of this
+# distribution for a copy of the license or check <https://www.gnu.org/licenses/>.
+#
+# SPDX-License-Identifier: GPL-3.0-or-later
+import numpy as np
+import pytest
+
+from eve import pattern_matching
+from functional.common import Field, GTTypeError
+from functional.ffront import common_types
+from functional.ffront import program_ast as past
+from functional.ffront.decorator import field_operator, program
+from functional.ffront.func_to_past import ProgramParser, ProgramSyntaxError
+from functional.ffront.past_to_itir import ProgramLowering
+from functional.iterator import ir as itir
+from functional.iterator.embedded import index_field, np_as_located_field
+from functional.iterator.runtime import CartesianAxis, offset
+
+
+float64 = float
+IDim = CartesianAxis("IDim")
+Ioff = offset("Ioff")
+
+past_ = pattern_matching.ModuleWrapper(past)
+itir_ = pattern_matching.ModuleWrapper(itir)
+
+
+@field_operator
+def identity(in_field: Field[[IDim], "float64"]) -> Field[[IDim], "float64"]:
+    return in_field
+
+
+@program
+def copy_program(in_field: Field[[IDim], "float64"], out_field: Field[[IDim], "float64"]):
+    identity(in_field, out=out_field)
+
+
+@program
+def double_copy_program(
+    in_field: Field[[IDim], "float64"],
+    intermediate_field: Field[[IDim], "float64"],
+    out_field: Field[[IDim], "float64"],
+):
+    identity(in_field, out=intermediate_field)
+    identity(intermediate_field, out=out_field)
+
+
+@program
+def copy_restrict_program(in_field: Field[[IDim], "float64"], out_field: Field[[IDim], "float64"]):
+    identity(in_field, out=out_field[1:2])
+
+
+def invalid_call_sig_program(
+    in_field: Field[[IDim], "float64"], out_field: Field[[IDim], "float64"]
+):
+    identity(in_field, in_field, out=out_field)
+
+
+def invalid_slice_program(in_field: Field[[IDim], "float64"], out_field: Field[[IDim], "float64"]):
+    identity(in_field, out=out_field[1:2, 3:4])
```

Comment:
You are absolutely right. All field operators are now inside a fixture, but decorated as this is required before being used inside a program (at least with respect to being executable). It is possible to avoid this, but not without duplicating programs for parsing+lowering and execution. Programs on the other hand are undecorated now and only decorated where needed. In total I must say the resulting fixture code doesn't feel elegant, so I am open for suggestions on improvement.